consensus/src/processes/difficulty.rs

```diff
@@ -48,20 +57,78 @@ impl<T: HeaderStoreReader> DifficultyManager<T> {
         (sp_daa_score + daa_added_blocks.len() as u64, daa_added_blocks)
     }
 
-    pub fn calculate_difficulty_bits(&self, window_hashes: &mut impl ExactSizeIterator<Item = Hash>) -> u32 {
+    pub fn calculate_difficulty_bits(&self, window: &BlockWindowHeap) -> u32 {
+        let mut difficulty_blocks: Vec<DifficultyBlock> = window
+            .iter()
+            .map(|item| {
+                let data = self
+                    .headers_store
+                    .get_compact_header_data(item.0.hash)
+                    .unwrap();
+                DifficultyBlock { timestamp: data.timestamp, bits: data.bits, sortable_block: item.0.clone() }
+            })
+            .collect();
+
         // Until there are enough blocks for a full block window the difficulty should remain constant.
-        if window_hashes.len() < self.difficulty_adjustment_window_size {
+        if difficulty_blocks.len() < self.difficulty_adjustment_window_size {
             return self.genesis_bits;
         }
 
-        let window_headers = window_hashes.map(|hash| self.headers_store.get_header(hash));
+        let (min_ts_index, min_diff_block) = difficulty_blocks
+            .iter()
+            .enumerate()
+            .min_by(|(_, a), (_, b)| a.cmp(b))
+            .unwrap();
+        let min_ts = min_diff_block.timestamp;
+        let max_diff_block = difficulty_blocks.iter().max().unwrap();
+        let max_ts = max_diff_block.timestamp;
+
+        let targets_len = difficulty_blocks.len() - 1;
+
+        // We remove the minimal block because we want the average target for the internal window.
+        difficulty_blocks.swap_remove(min_ts_index);
+
+        let targets = difficulty_blocks
+            .into_iter()
+            .map(|diff_block| compact_to_target(diff_block.bits));
+
+        let targets_sum: u64 = targets.sum();
+        let average_target: u64 = targets_sum / targets_len as u64;
```

Comment:
Maybe inline all to something like `avg = blocks.into_iter().map(...).sum() / targets_len`